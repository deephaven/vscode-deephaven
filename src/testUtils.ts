/**
 * This module contains test utils that should only be consumed by `.spec.ts`
 * files, but since it's not a `.spec.ts` file, it is special cased to exclude
 * from the extension build (tsconfig.json) and include in the unit test build
 * (tsconfig.unit.json). It shouldn't be imported by the extension code or have
 * any dependencies on the extension code.
 */

import { expect, vi } from 'vitest';
import type * as vscode from 'vscode';
export const bitValues = [0, 1] as const;
export const boolValues = [true, false] as const;

export type MessageHandler<TData> = (
  event: Partial<MessageEvent<TData>>
) => unknown;

/**
 * Return last registered event handler for 'message' event. Assumes that
 * `window.addEventListener` has been mocked or spied on.
 */
export function getLastMessageHandler<TData>(): MessageHandler<TData> {
  const messageCalls = vi
    .mocked(window.addEventListener<'message'>)
    .mock.calls.filter(([type]) => type === 'message');

  expect(messageCalls.length).toBeGreaterThan(0);

  return messageCalls.at(-1)![1] as MessageHandler<TData>;
}

/**
 * Generate a 2 dimensional array of all possible combinations of the given value
 * lists.
 *
 * e.g.
 * matrix([1, 2], ['a', 'b']) => [[1, 'a'], [1, 'b'], [2, 'a'], [2, 'b']]
 *
 * matrix([1, 2], ['a', 'b', 'c']) => [
 *  [1, 'a'], [1, 'b'], [1, 'c'],
 *  [2, 'a'], [2, 'b'], [2, 'c'],
 * ]
 *
 * @param args Value lists
 * @returns 2D array of all possible combinations
 */
export function matrix<
  TArgs extends (unknown[] | readonly unknown[])[],
  TReturn extends { [P in keyof TArgs]: TArgs[P][number] },
>(...args: TArgs): TReturn[] {
  const [first, ...rest] = args;

  if (rest.length === 0) {
    return first.map(value => [value] as TReturn);
  }

  // recursively call matrix
  const restMatrix = matrix(...rest);

  return first.flatMap(value =>
    restMatrix.map(values => [value, ...values] as TReturn)
  );
}

function lineAt(this: vscode.TextDocument, line: number): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  position: vscode.Position
): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  lineOrPosition: number | vscode.Position
): vscode.TextLine {
  const lineNumber =
    typeof lineOrPosition === 'number' ? lineOrPosition : lineOrPosition.line;

  return mockT<vscode.TextLine>({
    lineNumber,
    text: this.getText().split('\n')[lineNumber] ?? '',
  });
}

/** Mock a `vscode.TextDocument` */
export function mockDocument(
  fileName: string,
  version: number = 1,
  content = ''
): vscode.TextDocument {
  return mockT<vscode.TextDocument>({
    fileName,
    uri: mockUri(`file://mock/path/${fileName}`),
    version,
    getText: () => content,
    lineAt,
  });
}

/** Mock an object providing partial properties */
export function mockT<T>(partial: Partial<T>): T {
  return partial as T;
}

/** Mock a `vscode.Uri` */
export function mockUri(path: string): vscode.Uri {
  return mockT<vscode.Uri>({
    path,
    toString: () => path,
  });
}
