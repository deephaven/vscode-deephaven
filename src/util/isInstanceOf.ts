/**
 * Type guard wrapper of `instanceof` operator. This is not actually necessary
 * as a typeguard, since `instanceof` is already a typeguard, but it allows us
 * to mock `instanceof` checks in unit tests.
 * @param value The value to check
 * @param type The type to check against
 * @returns `true` if `value` is an instance of `type`
 */
export function isInstanceOf<T>(
  value: unknown,
  // Possible we'll need to include non-abstract constructor type, but this seems
  // to work for both abstract and non-abstract constructors.
  type: abstract new (...args: any[]) => T
): value is T {
  return value instanceof type;
}
