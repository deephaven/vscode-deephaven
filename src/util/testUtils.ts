import type * as vscode from 'vscode';
export const bitValues = [0, 1] as const;
export const boolValues = [true, false] as const;

/**
 * Generate a 2 dimensional array of all possible combinations of the given value
 * lists.
 *
 * e.g.
 * matrix([1, 2], ['a', 'b']) => [[1, 'a'], [1, 'b'], [2, 'a'], [2, 'b']]
 *
 * matrix([1, 2], ['a', 'b', 'c']) => [
 *  [1, 'a'], [1, 'b'], [1, 'c'],
 *  [2, 'a'], [2, 'b'], [2, 'c'],
 * ]
 *
 * @param args Value lists
 * @returns 2D array of all possible combinations
 */
export function matrix<
  TArgs extends (unknown[] | readonly unknown[])[],
  TReturn extends { [P in keyof TArgs]: TArgs[P][number] },
>(...args: TArgs): TReturn[] {
  const [first, ...rest] = args;

  if (rest.length === 0) {
    return first.map(value => [value] as TReturn);
  }

  // recursively call matrix
  const restMatrix = matrix(...rest);

  return first.flatMap(value =>
    restMatrix.map(values => [value, ...values] as TReturn)
  );
}

function lineAt(this: vscode.TextDocument, line: number): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  position: vscode.Position
): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  lineOrPosition: number | vscode.Position
): vscode.TextLine {
  const lineNumber =
    typeof lineOrPosition === 'number' ? lineOrPosition : lineOrPosition.line;

  return mockT<vscode.TextLine>({
    lineNumber,
    text: this.getText().split('\n')[lineNumber] ?? '',
  });
}

/** Mock a `vscode.TextDocument` */
export function mockDocument(
  fileName: string,
  version: number = 1,
  content = ''
): vscode.TextDocument {
  return mockT<vscode.TextDocument>({
    fileName,
    uri: mockUri(`file://mock/path/${fileName}`),
    version,
    getText: () => content,
    lineAt,
  });
}

/** Mock an object providing partial properties */
export function mockT<T>(partial: Partial<T>): T {
  return partial as T;
}

/** Mock a `vscode.Uri` */
export function mockUri(path: string): vscode.Uri {
  return mockT<vscode.Uri>({
    path,
    toString: () => path,
  });
}
