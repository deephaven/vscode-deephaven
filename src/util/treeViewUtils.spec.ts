import { describe, it, expect, vi } from 'vitest';
import { bitValues, boolValues, matrix } from './testUtils';
import {
  getPanelConnectionTreeItem,
  getPanelVariableTreeItem,
  getServerContextValue,
  getServerDescription,
  getServerGroupContextValue,
  getServerGroupTreeItem,
  getServerIconID,
  getServerTreeItem,
  getVariableIconPath,
  groupServers,
} from './treeViewUtils';
import type {
  ConsoleType,
  IDhService,
  ServerState,
  VariableDefintion,
  VariableType,
} from '../types';
import { isInstanceOf } from './isInstanceOf';

// See __mocks__/vscode.ts for the mock implementation
vi.mock('vscode');
vi.mock('../util/isInstanceOf.ts');

const variableTypes: readonly VariableType[] = [
  'deephaven.plot.express.DeephavenFigure',
  'deephaven.ui.Element',
  'Figure',
  'HierarchicalTable',
  'OtherWidget',
  'pandas.DataFrame',
  'PartitionedTable',
  'Table',
  'TableMap',
  'Treemap',
  'TreeTable',
] as const;

describe('getPanelConnectionTreeItem', () => {
  const getConsoleTypes: IDhService['getConsoleTypes'] = vi
    .fn()
    .mockResolvedValue(new Set<ConsoleType>(['python']));

  const serverUrl = new URL('http://localhost:10000');

  it.each(matrix(boolValues, boolValues))(
    'should return panel connection tree item: isConnected:%s, isInitialized:%s',
    async (isConnected, isInitialized) => {
      const connection = {
        isConnected,
        isInitialized,
        serverUrl,
        getConsoleTypes,
      } as IDhService;

      vi.mocked(isInstanceOf).mockReturnValue(true);

      const actual = await getPanelConnectionTreeItem(connection);
      expect(actual).toMatchSnapshot();
    }
  );
});

describe('getPanelVariableTreeItem', () => {
  const url = new URL('http://localhost:10000');

  it.each(variableTypes)(
    'should return panel variable tree item: type:%s',
    type => {
      const variable = {
        title: 'some title',
        type,
      } as VariableDefintion;

      const actual = getPanelVariableTreeItem([url, variable]);
      expect(actual).toMatchSnapshot();
    }
  );
});

describe('getServerContextValue', () => {
  it.each(matrix(boolValues, boolValues, boolValues))(
    'should return contextValue based on server state: isConnected=%s, isManaged=%s, isRunning=%s',
    (isConnected, isManaged, isRunning) => {
      const actual = getServerContextValue({
        isConnected,
        isDHE: false,
        isManaged,
        isRunning,
      });
      expect(actual).toMatchSnapshot();
    }
  );
});

describe('getServerDescription', () => {
  const labelValeus = ['some label', undefined] as const;

  it.each(matrix(bitValues, boolValues, labelValeus))(
    'should return server description based on parameters: connectionCount=%s, isManaged=%s, label=%s',
    (connectionCount, isManaged, label) => {
      const actual = getServerDescription(connectionCount, isManaged, label);
      expect(actual).toMatchSnapshot();
    }
  );
});

describe('getServerGroupContextValue', () => {
  const groupValues = ['Managed', 'Running'] as const;

  it.each(matrix(groupValues, boolValues))(
    'should return context value when servers can be managed: group=%s, canStartServer=%s',
    (group, canStartServer) => {
      const actual = getServerGroupContextValue(group, canStartServer);
      expect(actual).toMatchSnapshot();
    }
  );
});

describe('getServerGroupTreeItem', () => {
  const groupValues = ['Managed', 'Running'] as const;

  it.each(matrix(groupValues, boolValues))(
    'should return server group tree item: group=%s, canStartServer=%s',
    (group, canStartServer) => {
      const actual = getServerGroupTreeItem(group, canStartServer);
      expect(actual).toMatchSnapshot();
    }
  );
});

describe('getServerIconID', () => {
  it.each(matrix(boolValues, boolValues, boolValues))(
    'should return icon id based on server state: isConnected=%s, isManaged=%s, isRunning=%s',
    (isConnected, isManaged, isRunning) => {
      const actual = getServerIconID({ isConnected, isManaged, isRunning });
      expect(actual).toMatchSnapshot();
    }
  );
});

describe('getServerTreeItem', () => {
  const typeValues = ['DHC', 'DHE'] as const;

  const dhcServerState: ServerState = {
    type: 'DHC',
    url: new URL('http://localhost:10000'),
    isConnected: false,
    isRunning: false,
    connectionCount: 0,
  };

  it.each(matrix(typeValues, boolValues, boolValues, boolValues))(
    'should return server tree item: type=%s, isConnected=%s, isManaged=%s, isRunning=%s',
    (type, isConnected, isManaged, isRunning) => {
      const actual = getServerTreeItem({
        ...dhcServerState,
        ...(isManaged
          ? { isManaged: true, psk: 'mock.psk' }
          : { isManaged: false }),
        type,
        connectionCount: isConnected ? 1 : 0,
        isConnected,
        isRunning,
      });

      expect(actual).toMatchSnapshot();
    }
  );
});

describe('getVariableIconPath', () => {
  it('should return icon path for variableType', () => {
    expect(
      variableTypes.map(type => [type, getVariableIconPath(type)])
    ).toMatchSnapshot();
  });
});

describe('groupServers', () => {
  it('should group servers by state', () => {
    // Note that each combination is duplicated so that multiple servers get
    // created for each group.
    const props = matrix(boolValues, [true, true, false, false]);

    const servers = props.map(
      ([isManaged, isRunning], i) =>
        ({
          type: 'DHC' as const,
          url: new URL(`http://localhost:1000${i}`),
          isManaged,
          isRunning,
          psk: isManaged ? 'mock.psk' : undefined,
        }) as ServerState
    );

    const actual = groupServers(servers);

    expect(actual).toMatchSnapshot();
  });
});
