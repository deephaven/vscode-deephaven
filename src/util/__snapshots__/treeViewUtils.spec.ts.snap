// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getPanelConnectionTreeItem > should return panel connection tree item: isConnected:false, isInitialized:false 1`] = `
{
  "collapsibleState": 2,
  "description": undefined,
  "iconPath": {
    "color": undefined,
    "id": "sync~spin",
  },
  "label": "localhost:10000",
}
`;

exports[`getPanelConnectionTreeItem > should return panel connection tree item: isConnected:false, isInitialized:true 1`] = `
{
  "collapsibleState": 2,
  "description": "python",
  "iconPath": {
    "color": undefined,
    "id": "sync~spin",
  },
  "label": "localhost:10000",
}
`;

exports[`getPanelConnectionTreeItem > should return panel connection tree item: isConnected:true, isInitialized:false 1`] = `
{
  "collapsibleState": 2,
  "description": undefined,
  "iconPath": {
    "color": undefined,
    "id": "vm-connect",
  },
  "label": "localhost:10000",
}
`;

exports[`getPanelConnectionTreeItem > should return panel connection tree item: isConnected:true, isInitialized:true 1`] = `
{
  "collapsibleState": 2,
  "description": "python",
  "iconPath": {
    "color": undefined,
    "id": "vm-connect",
  },
  "label": "localhost:10000",
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:Figure 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "Figure",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "graph",
  },
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:HierarchicalTable 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "HierarchicalTable",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "dh-table",
  },
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:OtherWidget 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "OtherWidget",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "preview",
  },
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:PartitionedTable 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "PartitionedTable",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "dh-table",
  },
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:Table 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "Table",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "dh-table",
  },
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:TableMap 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "TableMap",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "dh-table",
  },
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:TreeTable 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "TreeTable",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "dh-table",
  },
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:Treemap 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "Treemap",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "preview",
  },
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:deephaven.plot.express.DeephavenFigure 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "deephaven.plot.express.DeephavenFigure",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "graph",
  },
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:deephaven.ui.Element 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "deephaven.ui.Element",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "preview",
  },
}
`;

exports[`getPanelVariableTreeItem > should return panel variable tree item: type:pandas.DataFrame 1`] = `
{
  "command": {
    "arguments": [
      "http://localhost:10000/",
      [
        {
          "title": "some title",
          "type": "pandas.DataFrame",
        },
      ],
    ],
    "command": "vscode-deephaven.openVariablePanels",
    "title": "Open Panel",
  },
  "description": "some title",
  "iconPath": {
    "color": undefined,
    "id": "dh-pandas",
  },
}
`;

exports[`getServerContextValue > should return contextValue based on server state: isConnected=false, isManaged=false, isRunning=false 1`] = `"isServerStopped"`;

exports[`getServerContextValue > should return contextValue based on server state: isConnected=false, isManaged=false, isRunning=true 1`] = `"isServerRunningDisconnected"`;

exports[`getServerContextValue > should return contextValue based on server state: isConnected=false, isManaged=true, isRunning=false 1`] = `"isManagedServerConnecting"`;

exports[`getServerContextValue > should return contextValue based on server state: isConnected=false, isManaged=true, isRunning=true 1`] = `"isManagedServerDisconnected"`;

exports[`getServerContextValue > should return contextValue based on server state: isConnected=true, isManaged=false, isRunning=false 1`] = `"isServerStopped"`;

exports[`getServerContextValue > should return contextValue based on server state: isConnected=true, isManaged=false, isRunning=true 1`] = `"isServerRunningConnected"`;

exports[`getServerContextValue > should return contextValue based on server state: isConnected=true, isManaged=true, isRunning=false 1`] = `"isManagedServerConnected"`;

exports[`getServerContextValue > should return contextValue based on server state: isConnected=true, isManaged=true, isRunning=true 1`] = `"isManagedServerConnected"`;

exports[`getServerDescription > should return server description based on parameters: connectionCount=0, isManaged=false, label=some label 1`] = `"some label"`;

exports[`getServerDescription > should return server description based on parameters: connectionCount=0, isManaged=false, label=undefined 1`] = `""`;

exports[`getServerDescription > should return server description based on parameters: connectionCount=0, isManaged=true, label=some label 1`] = `"pip some label"`;

exports[`getServerDescription > should return server description based on parameters: connectionCount=0, isManaged=true, label=undefined 1`] = `"pip"`;

exports[`getServerDescription > should return server description based on parameters: connectionCount=1, isManaged=false, label=some label 1`] = `"some label (1)"`;

exports[`getServerDescription > should return server description based on parameters: connectionCount=1, isManaged=false, label=undefined 1`] = `"(1)"`;

exports[`getServerDescription > should return server description based on parameters: connectionCount=1, isManaged=true, label=some label 1`] = `"pip some label (1)"`;

exports[`getServerDescription > should return server description based on parameters: connectionCount=1, isManaged=true, label=undefined 1`] = `"pip (1)"`;

exports[`getServerGroupContextValue > should return context value when servers can be managed: group=Managed, canStartServer=false 1`] = `undefined`;

exports[`getServerGroupContextValue > should return context value when servers can be managed: group=Managed, canStartServer=true 1`] = `"canStartServer"`;

exports[`getServerGroupContextValue > should return context value when servers can be managed: group=Running, canStartServer=false 1`] = `undefined`;

exports[`getServerGroupContextValue > should return context value when servers can be managed: group=Running, canStartServer=true 1`] = `undefined`;

exports[`getServerGroupTreeItem > should return server group tree item: group=Managed, canStartServer=false 1`] = `
{
  "collapsibleState": 2,
  "contextValue": undefined,
  "iconPath": {
    "color": undefined,
    "id": "server",
  },
  "label": "Managed",
}
`;

exports[`getServerGroupTreeItem > should return server group tree item: group=Managed, canStartServer=true 1`] = `
{
  "collapsibleState": 2,
  "contextValue": "canStartServer",
  "iconPath": {
    "color": undefined,
    "id": "server",
  },
  "label": "Managed",
}
`;

exports[`getServerGroupTreeItem > should return server group tree item: group=Running, canStartServer=false 1`] = `
{
  "collapsibleState": 2,
  "contextValue": undefined,
  "iconPath": {
    "color": undefined,
    "id": "server",
  },
  "label": "Running",
}
`;

exports[`getServerGroupTreeItem > should return server group tree item: group=Running, canStartServer=true 1`] = `
{
  "collapsibleState": 2,
  "contextValue": undefined,
  "iconPath": {
    "color": undefined,
    "id": "server",
  },
  "label": "Running",
}
`;

exports[`getServerIconID > should return icon id based on server state: isConnected=false, isManaged=false, isRunning=false 1`] = `"circle-slash"`;

exports[`getServerIconID > should return icon id based on server state: isConnected=false, isManaged=false, isRunning=true 1`] = `"circle-large-outline"`;

exports[`getServerIconID > should return icon id based on server state: isConnected=false, isManaged=true, isRunning=false 1`] = `"sync~spin"`;

exports[`getServerIconID > should return icon id based on server state: isConnected=false, isManaged=true, isRunning=true 1`] = `"circle-large-outline"`;

exports[`getServerIconID > should return icon id based on server state: isConnected=true, isManaged=false, isRunning=false 1`] = `"circle-slash"`;

exports[`getServerIconID > should return icon id based on server state: isConnected=true, isManaged=false, isRunning=true 1`] = `"circle-large-filled"`;

exports[`getServerIconID > should return icon id based on server state: isConnected=true, isManaged=true, isRunning=false 1`] = `"sync~spin"`;

exports[`getServerIconID > should return icon id based on server state: isConnected=true, isManaged=true, isRunning=true 1`] = `"circle-large-filled"`;

exports[`getServerTreeItem > should return server tree item: type=DHC, isConnected=false, isManaged=false, isRunning=false 1`] = `
{
  "command": undefined,
  "contextValue": "isServerStopped",
  "description": "",
  "iconPath": {
    "color": undefined,
    "id": "circle-slash",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHC, isConnected=false, isManaged=false, isRunning=true 1`] = `
{
  "command": {
    "arguments": [
      {
        "connectionCount": 0,
        "isConnected": false,
        "isManaged": false,
        "isRunning": true,
        "type": "DHC",
        "url": "http://localhost:10000/",
      },
    ],
    "command": "vscode-deephaven.connectToServer",
    "title": "Connect to server",
  },
  "contextValue": "isServerRunningDisconnected",
  "description": "",
  "iconPath": {
    "color": undefined,
    "id": "circle-large-outline",
  },
  "label": "localhost:10000",
  "tooltip": "Click to connect to http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHC, isConnected=false, isManaged=true, isRunning=false 1`] = `
{
  "command": undefined,
  "contextValue": "isManagedServerConnecting",
  "description": "pip",
  "iconPath": {
    "color": undefined,
    "id": "sync~spin",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHC, isConnected=false, isManaged=true, isRunning=true 1`] = `
{
  "command": {
    "arguments": [
      {
        "connectionCount": 0,
        "isConnected": false,
        "isManaged": true,
        "isRunning": true,
        "psk": "mock.psk",
        "type": "DHC",
        "url": "http://localhost:10000/",
      },
    ],
    "command": "vscode-deephaven.connectToServer",
    "title": "Connect to server",
  },
  "contextValue": "isManagedServerDisconnected",
  "description": "pip",
  "iconPath": {
    "color": undefined,
    "id": "circle-large-outline",
  },
  "label": "localhost:10000",
  "tooltip": "Click to connect to http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHC, isConnected=true, isManaged=false, isRunning=false 1`] = `
{
  "command": undefined,
  "contextValue": "isServerStopped",
  "description": "(1)",
  "iconPath": {
    "color": undefined,
    "id": "circle-slash",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHC, isConnected=true, isManaged=false, isRunning=true 1`] = `
{
  "command": undefined,
  "contextValue": "isServerRunningConnected",
  "description": "(1)",
  "iconPath": {
    "color": undefined,
    "id": "circle-large-filled",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHC, isConnected=true, isManaged=true, isRunning=false 1`] = `
{
  "command": undefined,
  "contextValue": "isManagedServerConnected",
  "description": "pip (1)",
  "iconPath": {
    "color": undefined,
    "id": "sync~spin",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHC, isConnected=true, isManaged=true, isRunning=true 1`] = `
{
  "command": undefined,
  "contextValue": "isManagedServerConnected",
  "description": "pip (1)",
  "iconPath": {
    "color": undefined,
    "id": "circle-large-filled",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHE, isConnected=false, isManaged=false, isRunning=false 1`] = `
{
  "command": undefined,
  "contextValue": "isServerStopped",
  "description": "",
  "iconPath": {
    "color": undefined,
    "id": "circle-slash",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHE, isConnected=false, isManaged=false, isRunning=true 1`] = `
{
  "command": {
    "arguments": [
      {
        "connectionCount": 0,
        "isConnected": false,
        "isManaged": false,
        "isRunning": true,
        "type": "DHE",
        "url": "http://localhost:10000/",
      },
    ],
    "command": "vscode-deephaven.connectToServer",
    "title": "Connect to server",
  },
  "contextValue": "isDHEServerRunning",
  "description": "",
  "iconPath": {
    "color": undefined,
    "id": "circle-large-outline",
  },
  "label": "localhost:10000",
  "tooltip": "Click to connect to http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHE, isConnected=false, isManaged=true, isRunning=false 1`] = `
{
  "command": undefined,
  "contextValue": "isManagedServerConnecting",
  "description": "pip",
  "iconPath": {
    "color": undefined,
    "id": "sync~spin",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHE, isConnected=false, isManaged=true, isRunning=true 1`] = `
{
  "command": {
    "arguments": [
      {
        "connectionCount": 0,
        "isConnected": false,
        "isManaged": true,
        "isRunning": true,
        "psk": "mock.psk",
        "type": "DHE",
        "url": "http://localhost:10000/",
      },
    ],
    "command": "vscode-deephaven.connectToServer",
    "title": "Connect to server",
  },
  "contextValue": "isManagedServerDisconnected",
  "description": "pip",
  "iconPath": {
    "color": undefined,
    "id": "circle-large-outline",
  },
  "label": "localhost:10000",
  "tooltip": "Click to connect to http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHE, isConnected=true, isManaged=false, isRunning=false 1`] = `
{
  "command": undefined,
  "contextValue": "isServerStopped",
  "description": "(1)",
  "iconPath": {
    "color": undefined,
    "id": "circle-slash",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHE, isConnected=true, isManaged=false, isRunning=true 1`] = `
{
  "command": {
    "arguments": [
      {
        "connectionCount": 1,
        "isConnected": true,
        "isManaged": false,
        "isRunning": true,
        "type": "DHE",
        "url": "http://localhost:10000/",
      },
    ],
    "command": "vscode-deephaven.connectToServer",
    "title": "Connect to server",
  },
  "contextValue": "isDHEServerRunning",
  "description": "(1)",
  "iconPath": {
    "color": undefined,
    "id": "circle-large-filled",
  },
  "label": "localhost:10000",
  "tooltip": "Click to connect to http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHE, isConnected=true, isManaged=true, isRunning=false 1`] = `
{
  "command": undefined,
  "contextValue": "isManagedServerConnected",
  "description": "pip (1)",
  "iconPath": {
    "color": undefined,
    "id": "sync~spin",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getServerTreeItem > should return server tree item: type=DHE, isConnected=true, isManaged=true, isRunning=true 1`] = `
{
  "command": undefined,
  "contextValue": "isManagedServerConnected",
  "description": "pip (1)",
  "iconPath": {
    "color": undefined,
    "id": "circle-large-filled",
  },
  "label": "localhost:10000",
  "tooltip": "http://localhost:10000/",
}
`;

exports[`getVariableIconPath > should return icon path for variableType 1`] = `
[
  [
    "deephaven.plot.express.DeephavenFigure",
    {
      "color": undefined,
      "id": "graph",
    },
  ],
  [
    "deephaven.ui.Element",
    {
      "color": undefined,
      "id": "preview",
    },
  ],
  [
    "Figure",
    {
      "color": undefined,
      "id": "graph",
    },
  ],
  [
    "HierarchicalTable",
    {
      "color": undefined,
      "id": "dh-table",
    },
  ],
  [
    "OtherWidget",
    {
      "color": undefined,
      "id": "preview",
    },
  ],
  [
    "pandas.DataFrame",
    {
      "color": undefined,
      "id": "dh-pandas",
    },
  ],
  [
    "PartitionedTable",
    {
      "color": undefined,
      "id": "dh-table",
    },
  ],
  [
    "Table",
    {
      "color": undefined,
      "id": "dh-table",
    },
  ],
  [
    "TableMap",
    {
      "color": undefined,
      "id": "dh-table",
    },
  ],
  [
    "Treemap",
    {
      "color": undefined,
      "id": "preview",
    },
  ],
  [
    "TreeTable",
    {
      "color": undefined,
      "id": "dh-table",
    },
  ],
]
`;

exports[`groupServers > should group servers by state 1`] = `
{
  "managed": [
    {
      "isManaged": true,
      "isRunning": true,
      "psk": "mock.psk",
      "type": "DHC",
      "url": "http://localhost:10000/",
    },
    {
      "isManaged": true,
      "isRunning": true,
      "psk": "mock.psk",
      "type": "DHC",
      "url": "http://localhost:10001/",
    },
    {
      "isManaged": true,
      "isRunning": false,
      "psk": "mock.psk",
      "type": "DHC",
      "url": "http://localhost:10002/",
    },
    {
      "isManaged": true,
      "isRunning": false,
      "psk": "mock.psk",
      "type": "DHC",
      "url": "http://localhost:10003/",
    },
  ],
  "running": [
    {
      "isManaged": false,
      "isRunning": true,
      "psk": undefined,
      "type": "DHC",
      "url": "http://localhost:10004/",
    },
    {
      "isManaged": false,
      "isRunning": true,
      "psk": undefined,
      "type": "DHC",
      "url": "http://localhost:10005/",
    },
  ],
  "stopped": [
    {
      "isManaged": false,
      "isRunning": false,
      "psk": undefined,
      "type": "DHC",
      "url": "http://localhost:10006/",
    },
    {
      "isManaged": false,
      "isRunning": false,
      "psk": undefined,
      "type": "DHC",
      "url": "http://localhost:10007/",
    },
  ],
}
`;
